### PR TITLE
Improved isPowerOf2.

### PR DESCRIPTION
**Pull request by TurkeyMan**
Less code, better codegen for signed case, passes unit tests. Removed the recursive instantiation.
I tried (but failed) to make the signed/unsigned cases the same code :(
I still suspect it's possible...

### CODE REVIEW

**Comment by tsbockman** [comment]
> Removed the recursive instantiation.

I made it recursive with `pragma(inline, true)` to prevent the compiler from emitting separate functions for sets that can use the same code, like `byte`, `short`, and `int`. I suppose it doesn't really matter though, as long as the inliner does its job.

File: std/math.d [hunk 1]
```diff
@@ -7723,21 +7723,14 @@ bool isPowerOf2(X)(const X x) pure @safe nothrow @nogc
...
-        else
-            alias W = typeof(x + 0u); // For ubyte and ushort W should be uint
-
-        static if (is(X == W))
         {
-            static if (isSigned!X)
-                return ((x & -x) == x) && (x > -x);
-            else
-                return (x & -x) > (x - 1);
+            auto y = cast(typeof(x + 0))x;
```
Comment (tsbockman) [inline std/math.d]: I'd prefer to see `const` instead of `auto` where possible. (No need to waste your time changing it here, though.)

**Comment by 9il** [comment]
Auto-merge toggled on

**Comment by TurkeyMan** [comment]
> to prevent the compiler from emitting separate functions for sets that can use the same code, like byte, short, and int.

Yeah, I'd be really disappointed to see this function actually emit a symbol... I should certainly hope this is always inlined.

**Comment by TurkeyMan** [comment]
> LGTM. Glad to see someone improve on my mediocre signed algorithm.

I'm still not sure this is optimal, I was just having trouble with the signed case being a lot worse than unsigned. Your unsigned test looked good, I couldn't think how to beat it while preserving correctness of the '0' case.

**Comment by tsbockman** [comment]
> I was just having trouble with the signed case being a lot worse than unsigned.

Just out of curiosity - what are you doing where you actually noticed this (let alone cared enough to fix it)? After all, my signed test was still really fast in absolute terms, though slower than the unsigned one.

**Comment by TurkeyMan** [comment]
Just reading the disassembly of my inner loop. There was a weird mess I couldn't understand appearing in the disasm, turns out it was the signed version of this function.
I was expecting to see the simplest version emit: `(x & (x-1)) == 0`, but then I realised this function handled the 0 case, which I don't care for, but I can see why (and your solution is nice, only one extra sub in the unsigned case). But even then it looked a lot longer than it needed to be. I think your signed version was aboout 6-7 bytes larger.
I couldn't find a way to do the signed version without a second comparison, so with that in mind, I figured the early-out for `<= 0` was best, since any call in that range does no work.

I still suspect it's possible to do the signed version with just one comparison though... ;)

**Pull request merged**